### PR TITLE
Add specification for UpcomingInvoice

### PR DESCRIPTION
**Pull request by codingshinoby**
## Summary
Add GetUpcomingInvoice endpoint for Storefront

### CODE REVIEW

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    description: |-
+      Number of times the upcoming invoice data has changed.
+
+      Use the revision number when analyzing webhook data to
+      determine if a change should take precedence over the current
+      representation.
+    type: integer
+  organizationId:
+    allOf:
+      - $ref: ./OrganizationId.yaml
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:36]: This shold be removed.

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    type: object
+    description: Embedded objects that are requested by the `expand` query parameter.
+    properties:
+      customer:
+        type: object
+      website:
+        type: object
+      organization:
+        type: object
+      leadSource:
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:199]: This should be removed.

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 1 above)
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:36]:
```suggestion
```

This fields should not be available to customer, thus removed, Storefront is customer API.

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        type: object
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:200]:
```suggestion
```

See https://github.com/Rebilly/api-definitions/pull/1748/files#r1462355279

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+          type: string
+          enum:
+            - self
+            - website
+            - customer
+            - organization
+            - subscription
+            - attachments
+            - transactionAllocations
+            - recalculateInvoice
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:188]:
```suggestion
            - self
```

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        type: string
+        description: Type of the tax ID number.
+        enum:
+          - eu-vat
+          - other
+        example: eu-vat
+      value:
+        type: string
+        description: Value of the tax ID number.
+        example: GB980780684
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:112]:
```suggestion
```

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    $ref: ./WebsiteId.yaml
+  subscriptionId:
+    description: ID of the related subscription order.
+    type: string
+    readOnly: true
+    maxLength: 50
+    example: ord_01GYJPRKHBD6ZYHH897QCJMBS4
+  currency:
+    x-sortable: true
+    x-basic: true
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:20]:
```suggestion
```

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    type: array
+    description: Upcoming invoice items array.
+    items:
+      $ref: ./StorefrontUpcomingInvoiceItem.yaml
+  amount:
+    description: Amount of the invoice.
+    type: number
+    x-type: Money
+    x-sortable: true
+    x-basic: true
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:47]:
```suggestion
```

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    format: double
+    readOnly: true
+  amountDue:
+    description: Amount that is due on the invoice.
+    type: number
+    x-type: Money
+    x-sortable: true
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:54]:
```suggestion
```

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+  subtotalAmount:
+    description: Subtotal amount of the invoice.
+    type: number
+    x-type: Money
+    format: double
+    readOnly: true
+  discountAmount:
+    description: Discount amount that is applied to the invoice.
+    type: number
+    x-type: Money
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:66]:
```suggestion
```

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+          format: double
+        description:
+          type: string
+          description: Description of the discount.
+        context:
+          $ref: ./DiscountContext.yaml
+  dueTime:
+    description: Date and time when the invoice is due for payment.
+    type: string
+    x-sortable: true
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:156]:
```suggestion
```

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+    format: date-time
+  issuedTime:
+    description: Date and time when the invoice is issued.
+    x-label: Date Issued
+    x-sortable: true
+    x-basic: true
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:162]:
```suggestion
```

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 1 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Done. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 4 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Done. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 5 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Done. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 9 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Done. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 11 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Done. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+    description: Amount that is due on the invoice.
+    type: number
+    x-type: Money
+    x-sortable: true
+    format: double
+    readOnly: true
+  subtotalAmount:
+    description: Subtotal amount of the invoice.
+    type: number
+    x-type: Money
```
[the remark this answers is not in the document]
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Done. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 8 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Done. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 10 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Done. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 7 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Done. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 6 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Done. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 3 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Done. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoiceItem.yaml [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    example: plan_0YV7DENSVGDBW9S71XZNNYYQ0X
+  subscriptionId:
+    type: string
+    description: ID of the order.
+    maxLength: 50
+    example: ord_01GYJPRKHBD6ZYHH897QCJMBS4
+  discountAmount:
+    description: Discount amount applied to the upcoming invoice item.
+    type: number
+    format: double
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoiceItem.yaml:49]: I think there is a mistake, there is not discountAmount in upcoming invoice items

Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoiceItem.yaml:49]:
```suggestion
```

File: openapi/components/schemas/StorefrontUpcomingInvoiceItem.yaml [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+      For example, an invoice item for a service is included in billing period number 3.
+      The invoice item is only applied to billing period number 3.
+    type: integer
+  createdTime:
+    $ref: ./CreatedTime.yaml
+  updatedTime:
+    $ref: ./UpdatedTime.yaml
+  tax:
+    description: Invoice item tax.
+    $ref: ./TaxItem.yaml
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoiceItem.yaml:70]:
```suggestion
```

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml [hunk 15]
```diff
@@ -0,0 +1,24 @@
+parameters:
+  - $ref: ../../components/parameters/resourceId.yaml
+get:
+  x-products:
+    - Storefront
+  tags:
+    - Storefront upcoming invoices
```
Comment (ashkarpetin) [inline openapi/paths/storefront/upcoming-invoices@{id}.yaml:7]:
```suggestion
    - Storefront orders
```

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+  summary: Retrieve a Upcoming Invoice
```
Comment (ashkarpetin) [inline openapi/paths/storefront/upcoming-invoices@{id}.yaml:8]:
```suggestion
  summary: Retrieve an upcoming subscription order invoice
```

File: openapi/openapi.yaml [hunk 17]
```diff
@@ -1327,6 +1327,8 @@ paths:
     $ref: paths/storefront/invoices.yaml
   '/storefront/invoices/{id}':
     $ref: paths/storefront/invoices@{id}.yaml
+  '/storefront/upcoming-invoices/{id}':
```
Comment (ashkarpetin) [inline openapi/openapi.yaml:1330]:
```suggestion
  '/storefront/orders/{id}/upcoming-invoice':
```

I am wondering why it is different from the Core?

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+  operationId: StorefrontGetUpcomingInvoice
```
Comment (ashkarpetin) [inline openapi/paths/storefront/upcoming-invoices@{id}.yaml:9]:
```suggestion
  operationId: StorefrontGetOrderUpcomingInvoice
```

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+  x-sdk-operation-name: get
+  security:
+    - CustomerJWT: []
+  description: Retrieves an upcoming invoice with a specified ID.
```
Comment (ashkarpetin) [inline openapi/paths/storefront/upcoming-invoices@{id}.yaml:13]:
```suggestion
  description: |-
    Retrieves an upcoming invoice for a specified subscription order.

    An upcoming invoice is an invoice that has not been issued.
    It functions as a preview of the next invoice for the order.
    For more information, see [Invoices](https://www.rebilly.com/docs/dev-docs/invoices/).
```

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+    200:
+      description: Upcoming invoice retrieved.
+      content:
+        application/json:
+          schema:
+            $ref: ../../components/schemas/StorefrontUpcomingInvoice.yaml
+    401:
+      $ref: ../../components/responses/Unauthorized.yaml
+    403:
+      $ref: ../../components/responses/Forbidden.yaml
```
Comment (ashkarpetin) [inline openapi/paths/storefront/upcoming-invoices@{id}.yaml:24]:
```suggestion
    403:
      $ref: ../../components/responses/Forbidden.yaml
    404:
      $ref: ../../components/responses/NotFound.yaml   
```

File: openapi/components/schemas/StorefrontUpcomingInvoiceItem.yaml (hunk 13 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoiceItem.yaml]: Understood. Removed it.

Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoiceItem.yaml]: Done. Thanks.

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml (hunk 15 above)
Comment (codingshinoby) [reply openapi/paths/storefront/upcoming-invoices@{id}.yaml]: Done. Thanks.

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml (hunk 16 above)
Comment (codingshinoby) [reply openapi/paths/storefront/upcoming-invoices@{id}.yaml]: Done. Thanks.

File: openapi/components/schemas/StorefrontUpcomingInvoiceItem.yaml (hunk 14 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoiceItem.yaml]: Done. Thanks.

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml (hunk 18 above)
Comment (codingshinoby) [reply openapi/paths/storefront/upcoming-invoices@{id}.yaml]: Done. Thanks.

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml (hunk 20 above)
Comment (codingshinoby) [reply openapi/paths/storefront/upcoming-invoices@{id}.yaml:28]: Done. Thanks.

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml (hunk 19 above)
Comment (codingshinoby) [reply openapi/paths/storefront/upcoming-invoices@{id}.yaml:13]: Done. Thanks.

File: openapi/openapi.yaml (hunk 17 above)
Comment (codingshinoby) [reply openapi/openapi.yaml]: Understood. Thanks!

File: openapi/openapi.yaml [hunk 21]
```diff
@@ -1327,6 +1327,8 @@ paths:
     $ref: paths/storefront/invoices.yaml
   '/storefront/invoices/{id}':
     $ref: paths/storefront/invoices@{id}.yaml
+  '/storefront/orders/{id}/upcoming-invoice':
+    $ref: paths/storefront/upcoming-invoices@{id}.yaml
```
Comment (ashkarpetin) [inline openapi/openapi.yaml:1331]: File should be renamed too

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+    readOnly: true
+    maxLength: 50
+    example: ord_01GYJPRKHBD6ZYHH897QCJMBS4
+  currency:
+    $ref: ./CurrencyCode.yaml
+  items:
+    type: array
+    description: Upcoming invoice items array.
+    items:
+      $ref: ./StorefrontUpcomingInvoiceItem.yaml
```
Comment (ashkarpetin) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:24]:
```suggestion
      $ref: ./UpcomingInvoiceItem.yaml
```

Perhaps to the same way as for StorefrontInvoice and use existing schema for items.

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+  summary: Retrieve an upcoming subscription order invoice
+  operationId: StorefrontGetOrderUpcomingInvoice
+  x-sdk-operation-name: get
+  security:
+    - CustomerJWT: []
+  description: Retrieves an upcoming invoice with a specified ID.
+  responses:
+    200:
+      description: |-
+        Retrieves an upcoming invoice for a specified subscription order.
```
Comment (ashkarpetin) [inline openapi/paths/storefront/upcoming-invoices@{id}.yaml:17]: Seems description placed in the wrong place, it should be in the core `description`

File: openapi/openapi.yaml (hunk 21 above)
Comment (codingshinoby) [reply openapi/openapi.yaml]: Renamed

File: openapi/paths/storefront/upcoming-invoices@{id}.yaml (hunk 23 above)
Comment (codingshinoby) [reply openapi/paths/storefront/upcoming-invoices@{id}.yaml]: Moved it. Thanks!

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml (hunk 22 above)
Comment (codingshinoby) [reply openapi/components/schemas/StorefrontUpcomingInvoice.yaml]: Changed. Thanks

File: openapi/paths/storefront/upcoming-invoice@{id}.yaml [hunk 24]
```diff
@@ -0,0 +1,29 @@
+parameters:
+  - $ref: ../../components/parameters/resourceId.yaml
+get:
+  x-products:
+    - Storefront
+  tags:
+    - Storefront orders
+  summary: Retrieve an upcoming subscription order invoice
+  operationId: StorefrontGetOrderUpcomingInvoice
+  x-sdk-operation-name: get
```
Comment (ashkarpetin) [inline openapi/paths/storefront/upcoming-invoice@{id}.yaml:10]:
```suggestion
  x-sdk-operation-name: getUpcomingInvoice
```

File: openapi/paths/storefront/upcoming-invoice@{id}.yaml [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+  security:
+    - CustomerJWT: []
+  description: |-
+    Retrieves an upcoming invoice for a specified subscription order.
+    An upcoming invoice is an invoice that has not been issued.
+    It functions as a preview of the next invoice for the order.
+    For more information, see [Invoices](https://www.rebilly.com/docs/dev-docs/invoices/).
```
Comment (ashkarpetin) [inline openapi/paths/storefront/upcoming-invoice@{id}.yaml:17]:
```suggestion
    Retrieves an upcoming invoice for a specified subscription order.
    
    An upcoming invoice is an invoice that has not been issued.
    It functions as a preview of the next invoice for the order.
    For more information, see [Invoices](https://www.rebilly.com/docs/dev-docs/invoices/).
```

File: openapi/paths/storefront/upcoming-invoice@{id}.yaml [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+  responses:
+    200:
+      content:
```
Comment (ashkarpetin) [inline openapi/paths/storefront/upcoming-invoice@{id}.yaml:20]:
```suggestion
    200:
      description: Upcoming invoice retrieved.
      content:
```

File: openapi/openapi.yaml [hunk 27]
```diff
@@ -1327,6 +1327,8 @@ paths:
     $ref: paths/storefront/invoices.yaml
   '/storefront/invoices/{id}':
     $ref: paths/storefront/invoices@{id}.yaml
+  '/storefront/orders/{id}/upcoming-invoice':
+    $ref: paths/storefront/upcoming-invoice@{id}.yaml
```
Comment (ashkarpetin) [inline openapi/openapi.yaml:1331]:
```suggestion
    $ref: paths/storefront/orders@{id}@upcoming-invoice.yaml
```

File: openapi/paths/storefront/upcoming-invoice@{id}.yaml (hunk 25 above)
Comment (codingshinoby) [reply openapi/paths/storefront/upcoming-invoice@{id}.yaml:17]: Fixed. Thanks!

File: openapi/openapi.yaml (hunk 27 above)
Comment (codingshinoby) [reply openapi/openapi.yaml:1331]: Fixed. Thanks!

**Review by ashkarpetin — APPROVED** [review]

**Review by noon-dawg — APPROVED** [review]
Looks good on the docs side

File: openapi/components/schemas/StorefrontUpcomingInvoice.yaml [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+properties:
+  id:
+    description: ID of the upcoming invoice, which is persisted in the future renewal invoice.
+    type: string
+    readOnly: true
+    maxLength: 50
+    example: in_0YVF9605RKC62BP14NE2R7V2XT
+  websiteId:
+    $ref: ./WebsiteId.yaml
+  subscriptionId:
```
Comment (creocoder) [inline openapi/components/schemas/StorefrontUpcomingInvoice.yaml:12]: We need remove that from specification. Upcoming invoice not related to any subscription anymore. Only in BC mode for legacy subscriptions.

File: openapi/paths/storefront/orders@{id}@upcoming-invoice.yaml [hunk 29]
```diff
@@ -0,0 +1,31 @@
...
+      content:
+        application/json:
+          schema:
+            $ref: ../../components/schemas/StorefrontUpcomingInvoice.yaml
+    401:
+      $ref: ../../components/responses/Unauthorized.yaml
+    403:
+      $ref: ../../components/responses/Forbidden.yaml
+    404:
+      $ref: ../../components/responses/NotFound.yaml
```
Comment (creocoder) [inline openapi/paths/storefront/orders@{id}@upcoming-invoice.yaml:31]: Description isn't correct there. Upcoming invoice isn't related to subscription anymore.

**Comment by creocoder** [comment]
@codingshinoby , @XaosSintez Please fix api specification and transformer for this resource.

**Pull request merged**